pkg/tests/tasks/security/certificate/external_ca_test.go

```diff
@@ -15,109 +15,193 @@
-			tmpDir, tmpDir)
-		shell.Executef(t,
-			`awk 'BEGIN {counter=0;} /BEGIN CERT/{counter++} { print > "%s/proxy-cert-" counter ".pem"}' < '%s/certs.pem'`,
-			tmpDir, tmpDir)
-
-		t.NewSubTest("root certificate").Run(func(t test.TestHelper) {
-			shell.Executef(t, `openssl x509 -in '%s' -text -noout > '%s/root-cert.crt.txt'`, sampleCARoot, tmpDir)
-			shell.Executef(t, `openssl x509 -in '%s/proxy-cert-3.pem' -text -noout > '%s/pod-root-cert.crt.txt'`, tmpDir, tmpDir)
-
-			if err := util.CompareFiles(fmt.Sprintf("%s/root-cert.crt.txt", tmpDir), fmt.Sprintf("%s/pod-root-cert.crt.txt", tmpDir)); err != nil {
-				t.Errorf("Root certs do not match: %v", err)
-			} else {
-				t.LogSuccess("Root certificate received from pod matches the root certificate in cacerts")
-			}
+		t.NewSubTest("Verify service certificate has trust chain to root certificate").Run(func(t test.TestHelper) {
+			verifyCertificate(t, certToVerify, []*x509.Certificate{rootCert}, otherCerts,
+				"Certificate trust chain successfully verified",
+				"Error: Unable to verify certificate trust chain")
 		})
+	})
+}
+
+func verifyContainsCerts(t test.TestHelper, certCollection []*x509.Certificate, certs []*x509.Certificate, successMsg, failureMsg string) {
+	t.T().Helper()
+
+	// Make a copy of certs because we will clobber it
+	certsToFind := make([]*x509.Certificate, len(certs), len(certs))
+	copy(certsToFind, certs)
+
+	for _, c1 := range certCollection {
+		for i2, c2 := range certsToFind {
+			if c1.Equal(c2) {
+				if len(certsToFind) == 1 {
+					t.LogSuccess(successMsg)
+					return
+				}
+				// "Remove" c2 from certsToFind
+				certsToFind[i2] = certsToFind[len(certsToFind)-1] // move last elem to where c2 was
+				certsToFind[len(certsToFind)-1] = nil             // remove old ref to last elem
+				certsToFind = certsToFind[:len(certsToFind)-1]    // resize slice
+				break
+			}
+		}
+	}
+	t.Error(failureMsg)
+}
 
-		t.NewSubTest("CA certificate").Run(func(t test.TestHelper) {
-			shell.Executef(t, `openssl x509 -in '%s' -text -noout > '%s/ca-cert.crt.txt'`, sampleCACert, tmpDir)
-			shell.Executef(t, `openssl x509 -in '%s/proxy-cert-2.pem' -text -noout > '%s/pod-cert-chain-ca.crt.txt'`, tmpDir, tmpDir)
+func verifyCertificate(t test.TestHelper, cert *x509.Certificate, rootCerts []*x509.Certificate, intermediateCerts []*x509.Certificate, successMsg, failureMsg string) {
+	t.T().Helper()
+	rootCertPool := x509.NewCertPool()
+	for _, c := range rootCerts {
+		rootCertPool.AddCert(c)
+	}
+
+	intermediateCertPool := x509.NewCertPool()
+	for _, c := range intermediateCerts {
+		intermediateCertPool.AddCert(c)
+	}
+
+	chains, err := cert.Verify(x509.VerifyOptions{
+		Roots:         rootCertPool,
+		Intermediates: intermediateCertPool})
+
+	var isUnkAuthErr = false
+	if err != nil {
+		_, isUnkAuthErr = err.(x509.UnknownAuthorityError)
+		if !isUnkAuthErr {
+			t.Fatalf(`Error while verifying certificate: %v`, err)
+		}
+	}
+
+	if !isUnkAuthErr && len(chains) > 0 {
```

Comment:
Do you need to check `isUnkAuthErr` here? Isn't it always false, since you call `t.Fatalf()` a few lines above, which causes the test to terminate?

Reply:
Yup it was completely redundant.